### DISCUSSION TITLE
How to Add Input in Panel Extension?

### DISCUSSION THREAD

**Discussion by ghost** [mecha-cms/mecha#99] [Q&A]
Saya baru saja mencoba install ekstensi panel.

Di ekstensi **panel**, saya menemukan formulir berisi `input` `title`, `slug`, `content`, `deskripsi`, dan opsi tipe `markdown` dan `HTML`.

Ketika saya membaca diskusi tentang [](https://mecha-cms.com/article/version-2-2-0?comments=1#comment:1614320440), kita bisa membuat `YAML` secara bebas, misalnya, `title_h1`.

Yang saya ingin tanyakan, bisakah saya menambah formulir isian di panel, semisal `heading 1`. Dan nantinya akan otomatis ada di data `YAML`?

Dari mana saya harus memulainya?

**Comment by taufik-nurrohman** [comment]
Ini pembahasannya bakalan panjang. Sekalian Saya bikin draf dokumentasi ekstensi panel saja di sini karena sampai sekarang memang masih belum ada.

Semua data panel tersedia sebagai _array_ statis, tersimpan secara global di dalam variabel `$_`. Untuk mengakses data terdapat banyak cara, tapi yang paling aman adalah melalui kait untuk memastikan agar semua data yang dimuat oleh ekstensi-ekstensi lain sudah tersedia:

~~~ php
Hook::set('_', function($_) {
    test($_);
    exit;
});
~~~

Beberapa data penting yang terkait secara langsung dengan variabel `$_`:

 - `f` &rarr; Berisi jalur ke berkas atau folder saat itu.
 - `form` &rarr; Berisi tipe permintaan (_request_) beserta datanya.
 `i` &rarr; Berisi angka paginasi saat itu jika ada.
 - `lot` &rarr; Berisi data GUI.
 - `path` &rarr; Jalur setelah `task`.
 - `task` &rarr; Tipe rute formulir. Secara umum nilainya bisa berupa `f`, `g`, `l`, atau `s`.
 - `title` &rarr; Berisi judul panel.
 - `type` &rarr; Berisi tipe panel.

Data GUI terdapat di dalam `$_['lot']`. Seharusnya terdiri dari dua bagian yaitu `bar` dan `desk`. `bar` berisi data menu dan `desk` berisi data konten, termasuk di dalam konten terdapat formulir, seksi dan tabulasi. Di dalam tabulasi kadang terdapat bidang-bidang formulir. Secara umum seperti itu, tapi pengembang bisa mengubahnya menjadi apa saja karena sifat data GUI yang sangat mentah. Saya sengaja tidak fokus ke pembuatan metode-metode untuk menambahkan atau mengurangi komponen-komponen di dalam panel, karena harapannya metode-metode tersebut nanti bisa dibuat oleh pengembang lain sebagai ekstensi terpisah. Sebagaimana fitur [ACF](https://www.advancedcustomfields.com/resources/#field-types/) pada WordPress.

~~~ yaml
lot:
  # Bagian menu
  bar: []
  # Bagian konten
  desk:
    lot:
      form:
        lot:
          # Bagian kepala
          # Biasanya berisi tajuk dan tombol-tombol aksi utama
          0: []
          # Bagian badan formulir
          1:
            lot:
              tabs:
                lot:
                  # Daftar tab
                  page: []
                  data: []
                  art: []
          # Bagian kaki
          # Biasanya berisi paginasi pada halaman indeks
          # Atau berisi tombol-tombol aksi formulir pada halaman item
          2: []
~~~

Pada daftar tab terdapat `fields` yang berisi bidang-bidang masukan untuk formulir:

~~~ yaml
page:
  lot:
    fields:
      lot:
        title: []
        content: []
        description: []
        type: []
data:
  lot:
    fields:
      lot:
        link: []
        time: []
art:
  lot:
    fields:
      lot:
        css: []
        js: []
~~~

Nama-nama _key_ pada setiap bidang tidak terlalu penting. Bagaimana data tersebut dikonversi nantinya akan ditentukan oleh bidang `type`. Nama-nama tipe bidang bisa dilihat [di sini](https://github.com/mecha-cms/x.panel.extra). Sedangkan bagaimana data tersebut diproses oleh formulir akan ditentukan oleh bidang `name`:

~~~ yaml
title:
  type: 'field/title'
  name: 'page[title]'
content:
  type: 'field/source'
  name: 'page[content]'
type:
  type: 'field/item'
  name: 'page[type]'
  lot:
    HTML: HTML
    Markdown: Markdown
# ...
time:
  type: 'field/date-time'
  name: 'data[time]'
~~~

Secara garis besar, struktur formulir yang dibentuk nantinya akan tampak seperti ini:

~~~ html
<form method="post">
  <input name="page[title]" type="text">
  <textarea name="page[content]"></textarea>
  <textarea name="page[description]"></textarea>
  <input name="page[type]" type="radio" value="HTML">
  <input name="page[type]" type="radio" value="Markdown">
  <!-- ... -->
  <input name="data[time]" type="text">
</form>
~~~

Ketika dikirim, data-data tersebut akan terkait ke `$_['form']['lot']`:

~~~ php
$_['form'] = [
    'lot' => [
        'data' => [
            'time' => ""
        ],
        'page' => [
            'content' => "",
            'description' => "",
            'title' => "",
            'type' => ""
        ],
        'token' => ""
    ],
    'type' => 'post'
];
~~~

`$_['form']['lot']['page']` akan diproses sebagai data berkas laman seperti ini:

~~~ yaml
---
title: ""
description: ""
type: ""
...

# Data `$_['form']['lot']['page']['content']` akan muncul di sini.
~~~

`$_['form']['lot']['data']` akan diproses secara terpisah sebagai data eksternal. `$_['form']['lot']['data']['time']` akan tersimpan sebagai berkas `time.data`, `$_['form']['lot']['data']['view']` akan tersimpan sebagai berkas `view.data`, dan seterusnya.

Untuk menambahkan bidang baru cukup kompleks karena ketika kita menambahkan bidang tanpa mengecek tipe panel terlebih dahulu, kadang akan menimbulkan efek-efek yang tidak diinginkan misalnya bidang yang ditambahkan tersebut malah jadi muncul di tipe-tipe panel yang bukan merupakan formulir. Jadi pertama-tama kita perlu menentukan kondisi yang tepat untuk menambahkan bidang baru:

~~~ php
Hook::set('_', function($_) {
    if (
        isset($_['lot']['desk']['lot']['form']['lot'][1]['lot']['tabs']['lot']['data']['lot']['fields']['lot']) &&
        0 === strpos($_['type'] . '/', 'page/')
    ) {
        // Tambahkan bidang baru di sini dengan aman
        if ($_['f']) {
            $page = new Page($_['f']);
        } else {
            $page = new Page;
        }
        $_['lot']['desk']['lot']['form']['lot'][1]['lot']['tabs']['lot']['data']['lot']['fields']['lot']['title-custom'] = [
            'title' => 'Heading',
            'description' => 'Title for heading.',
            'type' => 'field/text', // atau `text` saja
            'name' => 'page[title_h1]',
            'value' => $page['title_h1'],
            'stack' => 40
        ];
    }
    return $_;
});
~~~

**Reply by igoynawamreh** [reply]
@icchaahyuniswari  coba lihat [contoh ini](https://github.com/igoynawamreh/mecha101/blob/master/lot/layout/index/x.panel/page/editor/_about.php).

Kode dan strukturnya masih berantakan, nanti kalau ada waktu akan dibuat ulang, dibimbing mas @taufik-nurrohman 😛

**Reply by igoynawamreh** [reply]
Ada rencana bikin ekstensi untuk membuat bidang di panel dengan YAML gak mas?


Misal namanya `panel.blueprints`.

**Reply by taufik-nurrohman** [reply]
Itu bagianmu mas. Tinggal buka berkas YAML terus ubah jadi _array_, terus digabungkan ke `$GLOBALS['_']`:

~~~ .php
$data = From::YAML(content('.\path\to\panel-definitions.yaml') ?? '[]');

$GLOBALS['_'] = extend($GLOBALS['_'], (array) $data);

**Reply by yuliakbard** [reply]
cara memanggilnya yang nilai value `$page['title_h1']` itukah kang?

**Reply by taufik-nurrohman** [reply]
Iya. Lebih bagus lagi pakai akses objek: `$page->title_h1`.

**Comment by ghost** [comment]
Saya mencoba menambah isian formulir di panel.

Di dalam berkas `C:\xampp\htdocs\mecha-main\lot\x\panel\engine\r\type\page.php` saya tambahkan begini:

```php
// ...
'headline' => [
    'type' => 'headline',
    'name' => 'page[headline]',
    'value' => $page['headline'],
    'width' => true,
'stack' => 40
],
// ...
```

Dan, di folder `C:\xampp\htdocs\mecha-main\lot\x\panel\engine\f\type\field.php` saya juga membuat script ini dengan meniru `title`:

```php
function headline($value, $key) {
    $v = (string) ($value['value'] ?? "");
    if (!\array_key_exists('hint', $value)) {
        $value['hint'] = "" !== $v ? $v : 'Headline Goes Here';
    }
    if (!isset($value['max'])) {
        $value['max'] = 200;
    }
    return \x\panel\type\field\text($value, $key);
}
```

Saya cek dan lakukan percobaan submit, ternyata berhasil. Di laman terkait, nilai `headline` dalam `YAML`.

Apakah langkah saya sudah tepat?

Saya lihat ini juga mengubah semua formulir `comment`, `page`, `tag`, `user`. Semuanya memiliki isian `headline`, padahal saya hanya ingin menambahkan isian `headline` hanya di `page` khusus posting artikel saja. Meski ini tidak menjadi masalah karena setiap formulir tidak memiliki atribut `required`, dan jika tidak diisi tidak lantas membuat YAML memiliki `headline` dengan nilai kosong.

**Reply by taufik-nurrohman** [reply]
> Apakah langkah saya sudah tepat?

Konsepnya sudah tepat tapi akan lebih ideal kalau itu semua dilakukan di luar ekstensi _core_ panel (yaitu dengan cara membuat ekstensi baru untuk panel).

> Saya lihat ini juga mengubah semua formulir `comment`, `page`, `tag`, `user`. Semuanya memiliki isian _headline_, padahal saya hanya ingin menambahkan isian _headline_ hanya di page khusus posting artikel saja.

Itu karena tipe `page` terlalu generik. Berfungsi sebagai formulir untuk menangani berkas-berkas dengan ekstensi `.archive`, `.draft` dan `.page`.

Kalau mau lebih spesifik bisa dengan mengecek kondisi tipe sebagai `page`, `page/comment`, `page/page`, `page/tag`, atau `page/user`. Formulir khusus untuk posting artikel ada sebagai `page/page` dan datanya tersimpan di `.\lot\x\panel\engine\r\type\page\page.php`.

**Comment by yuliakbard** [comment]
Panel input type field/ option, setelah kita lakukan update postingan, option yang kita pilih itu selalu reset ke semula. saat kita kan pengen check ulang postingan/ update/ edit, option yang kita pilih sebelumnya itu, tampilnya tidak tetap di panel, jadi misalnya data yang sudah kita masukkan sebelum nya sudah benar, saat kita lakukan update, datanya berubah. bagaimana cara kita mengakalinya kang? berikut contoh panel field option yang saya gunakan

``` php
        $_['lot']['desk']['lot']['form']['lot'][1]['lot']['tabs']['lot']['jobPosting']['lot']['fields']['lot']['type_salary'] = [
            'title' => 'Tipe Gaji',
            'description' => 'Harian/ Mingguan/ Bulanan.',
            'type' => 'field/option',
            'name' => 'page[type_salary]',
            'value' => $page['type_salary'],
            'stack' => 22,
            'lot' => [
                0 => [
                    'title' => 'DAY',
                    'value' => 'DAY'
                    ],
                1 => [
                    'title' => 'WEEK',
                    'value' => 'WEEK'
                    ],
                2 => [
                    'title' => 'MONTH',
                    'value' => 'MONTH'
                    ],
                3 => [
                    'title' => 'YEAR',
                    'value' => 'YEAR'
                    ],
                4 => [
                    'title' => 'HOUR',
                    'value' => 'HOUR'
                    ]
                ]
        ];
```

**Reply by taufik-nurrohman** [reply]
Pas data kesimpen, bagian properti `type_salary` jadinya seperti apa?

**Reply by lokersumatera** [reply]
Yang ini done, penulisannya seperti ini. jadi valuenya tersimpan dengan tepat saat digunakan

``` php
        $_['lot']['desk']['lot']['form']['lot'][1]['lot']['tabs']['lot']['jobPosting']['lot']['fields']['lot']['type_salary'] = [
            'title' => 'Tipe Gaji',
            'description' => 'Harian/ Mingguan/ Bulanan.',
            'type' => 'field/option',
            'name' => 'page[type_salary]',
            'value' => $page['type_salary'],
            'stack' => 22,
            'lot' => [
                        'MONTH'=>'Per Bulan',
                        'WEEK'=>'Per Minggu',
                        'DAY'=>'Per Hari',
                        'YEAR'=>'Per Tahun',
                        'HOUR'=>'Per Jam'
            ]
        ];
```

**Comment by yuliakbard** [comment]
di mecha 3.0.0 ini, kode ini error kang.

```
Hook::set('_', function($_) {
    if (
        isset($_['lot']['desk']['lot']['form']['lot'][1]['lot']['tabs']['lot']['data']['lot']['fields']['lot']) &&
        0 === strpos($_['type'] . '/', 'page/')
    ) {
        // Tambahkan bidang baru di sini dengan aman
        if ($_['f']) {
            $page = new Page($_['f']);
        } else {
            $page = new Page;
        }
        $_['lot']['desk']['lot']['form']['lot'][1]['lot']['tabs']['lot']['data']['lot']['fields']['lot']['title-custom'] = [
            'title' => 'Heading',
            'description' => 'Title for heading.',
            'type' => 'field/text', // atau `text` saja
            'name' => 'page[title_h1]',
            'value' => $page['title_h1'],
            'stack' => 40
        ];
    }
    return $_;
});
```

**Reply by taufik-nurrohman** [reply]
Ganti `$_['f']` jadi `$_['file']`.

**Comment by yuliakbard** [comment]
Cara menampilkan item panel  `'type' => 'field/items',` gimana ya kang? kalau ane panggil seperti biasa `$page->type_tamatan` outputnya `Array`

```
$_['lot']['desk']['lot']['form']['lot'][1]['lot']['tabs']['lot']['page']['lot']['fields']['lot']['type_tamatan'] = [
            'title' => 'Tamatan',
            'description' => 'SD/ SMP/ SMA/ SMK/ D3/ S1 / D4/ S2 / S3.',
            'type' => 'field/items',
            'name' => 'page[type_tamatan]',
            'value' => $page['type_tamatan'],
            'stack' => 2.3,
            'lot' => [
            'SD'=>'SD',
            'SMP'=>'SMP',
            'SMA'=>'SMA',
            'SMK'=>'SMK',
            'D3'=>'D3',
            'S1'=>'S1',
            'D4'=>'D4',
            'S2'=>'S2',
            'S3'=>'S3',
            ]
        ];
```

**Reply by taufik-nurrohman** [reply]
Array-nya kaya apa?

**Reply by yuliakbard** [reply]
Kalau ane tulis `<?= $page->type_tamatan; ?>` outputnya tulisan `Array` , maunya itu outputnya sesuai dengan item yang ane select di panel, contoh output yang di harapkan: `SD, SMP` . atau kode input panel ane ada yang salah penulisannya kang?

**Reply by taufik-nurrohman** [reply]
Kalau begitu pakainya `field/item`, jangan `field/items`.

**Reply by yuliakbard** [reply]
kalau `field/item` cuma tampil satu aja kang, ini dia beberapa. kalau di looping gimana ambil items nya kang?

```
        <?php foreach($page->type_tamatan as $tamatan):?>
        <?=$tamatan->...  ;?>
        <?php endforeach;?>
```

**Reply by yuliakbard** [reply]
kalau di looping ambil item di dalam lot nya itu gimana ya kang?

**Reply by taufik-nurrohman** [reply]
`<?=$tamatan;?>`
